src/sentry/issues/auto_source_code_config/in_app_stack_trace_rules.py

```diff
@@ -0,0 +1,63 @@
+import logging
+from collections.abc import Mapping, Sequence
+from typing import Any
+
+from sentry.issues.auto_source_code_config.code_mapping import CodeMapping
+from sentry.issues.auto_source_code_config.utils import PlatformConfig
+from sentry.models.project import Project
+from sentry.utils import metrics
+
+from .constants import METRIC_PREFIX
+
+logger = logging.getLogger(__name__)
+
+OPTION_KEY = "sentry:automatic_grouping_enhancements"
+
+
+def save_in_app_stack_trace_rules(
+    project: Project, code_mappings: Sequence[CodeMapping], platform_config: PlatformConfig
+) -> list[str]:
+    """Save in app stack trace rules for a given project"""
+    tags: Mapping[str, str | bool] = {
+        "platform": platform_config.platform,
+        "dry_run": platform_config.is_dry_run_platform(),
+    }
+    in_app_stack_trace_rules = generate_in_app_stack_trace_rules(code_mappings)
+    if not platform_config.is_dry_run_platform():
+        project.update_option(OPTION_KEY, "\n".join(in_app_stack_trace_rules))
+
+    metrics.incr(
+        key=f"{METRIC_PREFIX}.in_app_stack_trace_rules.created",
+        amount=len(in_app_stack_trace_rules),
+        tags=tags,
+        sample_rate=1.0,
+    )
+    return in_app_stack_trace_rules
+
+
+def generate_in_app_stack_trace_rules(code_mappings: Sequence[CodeMapping]) -> list[Any]:
+    """Generate in app stack trace rules for a given project"""
+    rules = set()
+    for code_mapping in code_mappings:
+        rules.add(generate_rule_for_code_mapping(code_mapping))
+    return list(rules)
+
+
+# XXX: This is very Java specific. If we want to support other languages, we need to
+# come up with a better way to generate the rule.
+def generate_rule_for_code_mapping(code_mapping: CodeMapping) -> str:
+    """Generate a rule for a given code mapping"""
+    stacktrace_root = code_mapping.stacktrace_root
+    # Remove trailing slash
+    if stacktrace_root[-1] == "/":
+        stacktrace_root = stacktrace_root[:-1]
+
+    # For packages including the domain name, we want to have less granularity to create less rules, thus,
```

Comment:
```suggestion
    # For packages including the domain name, we want to have less granularity to create fewer rules, thus
```

English is ~fussy~ fun. 😛